### ISSUE TITLE
Detect async When calls

### ISSUE THREAD

**Issue by tpodolak** [nsubstitute/NSubstitute.Analyzers#81]
https://github.com/nsubstitute/NSubstitute/issues/467#issuecomment-431494787

**Comment by tpodolak** [comment]
@zvirja @dtchepak as mentioned in https://github.com/nsubstitute/NSubstitute/issues/467#issuecomment-431494787 we should not await in When callback. @zvirja's fix will be available in the next release of NSubstitute, so how analyzers should behave when someone awaits When callback and uses older version of NSubstitute? Should I suggest not using await (which will result in CS4014: Async method invocation without await expression), or should I suggest to upgrade NSubstitute or maybe sth else?

**Comment by dtchepak** [comment]
Are you worried about people fixing that error then getting a compilation warning about not using await? In that case I think it is reasonable to suggest not using `await` and ensure they are using NSubstitute 4.x or above. (where `x` is whatever the next release ends up being :) ).

@zvirja ?

**Comment by zvirja** [comment]
@tpodolak @dtchepak As to me, I'd say that if people don't update `NSubstitute`, it's unlikely they update `NSubstitute.Analyzers` package 😅 Therefore, I find it a rare case when people have new analyzers while still using old version of NSubstitute.

As for the analyzer, I'd say that it should suggest a fix to use right API _only_ if that API is available. If all of a sudden customer uses older version of NSubstitute - he probably has a reason and we shouldn't annoy him with suggestion to update every time.

**Comment by tpodolak** [comment]
@dtchepak @zvirja thanks for the feedback. I will wait for next NSubstitute release before starting working on this one. Do we have any "release item" in GitHub so I can track it?

**Comment by dtchepak** [comment]
@tpodolak I don't think our release procedure is that organised. 😭 @zvirja do you have any ideas about this?

**Comment by zvirja** [comment]
No, I'm afraid we don't have. It's only my head, where I keep tracking this I would like to improve till the next release. Probably, I should be more transparent and raise them as issues...

**Comment by tpodolak** [comment]
@zvirja @dtchepak as stated in the original ticket 
````c#
substitute.When(async x => await x.Bar());
````
usages are not allowed and we should use
````c#
substitute.When(x => x.Bar());
````
what about this situation
````c#
public void Test1()
{
    var substitute = Substitute.For<IFoo>();
    substitute.When(LocalBar);
}

private static async Task LocalBar(IFoo x)
{
    await x.Bar();
}
````
is this considered as awaited `When` callback?

**Comment by dtchepak** [comment]
@zvirja Do you know the answer to this? I haven't done much with `async`/`await`. :-\

**Comment by zvirja** [comment]
@tpodolak Yes, sure. Doesn't it fail if you run the code? In fact, it's just another variation of the original sample - here the lambda body is just explicit.

I'm just curious whether it's possible to detect such kind of issues and it worth investing resources into that. You can always make a weird case which is not recognized, so we should probably take care of basic scenarios only... Anyway, you know spirit of the analyzers better, so it's up to you ;-)

**Comment by zvirja** [comment]
@dtchepak @tpodolak I tried to play with it a bit more and found that I cannot somehow reproduce the original issue anymore. I even switched to the original v3 branch - it simply doesn't happen 😕 

No matter whether I put `await` or not - it's not reproducible. That's because even though the state machine is generated, exception is captured to the returned `Task` object, which is never awaited. This way the failure is simply ignored.

I think it still makes sense to suggest overload without `Task`, but I don't see scenario anymore when it's critically important to do that.

@dtchepak Given that you managed somehow reproduce the original scenario, could you please elaborate on this? 😉

**Comment by tpodolak** [comment]
@dtchepak @zvirja I also cannot reproduce the issue - it works for me with and without await when running against NSubstitute 3.1.0, 4.0 and 4.1 (running on Ubuntu 18.04.1 with dotnet  2.2.203)

**Comment by dtchepak** [comment]
@tpodolak @zvirja I can't reproduce it anymore either. 🤦‍♂ I'm now running VS for Mac 8.1 with dotnet 2.2.203. I tried targeting dotnet core 2.0 as per my original repro comment, but I couldn't get the test to fail.

**Comment by zvirja** [comment]
Awesome 🤣

So should we add analyzers for this case? As an engineer I can understand the intention of doing that, but given that it brings no value - does it worth it? Especially, when people will start asking "what should I apply this change for?" 🤔

**Comment by tpodolak** [comment]
I would vote for not adding the analyzer in that case - less code to maintain and smaller memory footprint of analyzers overall. If in the future we will figure out the real cause I an always restore the branch (as the code is 90% ready)

**Comment by dtchepak** [comment]
I agree with @tpodolak . Sorry for the heisenissue!

**Comment by tpodolak** [comment]
Closing as discussed

**Comment by TobiasWenzel** [comment]
Actually I just stumbled across this issue because I put a `When()` with an `async/await` handler into it. But not because I got a compiler warning when not doing it, but just because I thought it would be the correct way. 😉
Although it did not seem to do any harm, I wanted to get rid of that `NullReferenceException` popping up every time in Visual Studio. Then I came across this Github issue.

For me, an implementation of that analyzer would have been beneficial. But, as a software developer, I also see the "less code to maintain" point... 😉
